Use named constructors to construct String, StringView from Span objects

zdobersek opened the pull request:
#### 1a8542ef43120d556721c573faff4085ea0446f3
<pre>
Use named constructors to construct String, StringView from Span objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=251385">https://bugs.webkit.org/show_bug.cgi?id=251385</a>

Reviewed by NOBODY (OOPS!).

Turn conversion constructors on String and StringView classes accepting Span
objects into named constructor equivalents. This avoids unintentional use of
constructors through implicit construction of Span objects from e.g. containers
like Vectors, C arrays or string literals.

For latin-1 data spans, fromLatin1Span() is provided. For UTF-16 data spans, the
generic fromSpan() is provided. StringView constructors are regrouped and
reordered to mirror grouping and ordering of String constructors.

* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::canonicalizeTimeZoneName):
(JSC::IntlDateTimeFormat::format const):
(JSC::IntlDateTimeFormat::createDateIntervalFormatIfNecessary):
(JSC::IntlDateTimeFormat::formatRange):
* Source/JavaScriptCore/runtime/IntlDisplayNames.cpp:
(JSC::IntlDisplayNames::of const):
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::collectElements):
(JSC::IntlDurationFormat::format const):
* Source/JavaScriptCore/runtime/IntlListFormat.cpp:
(JSC::IntlListFormat::format const):
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::format const):
(JSC::IntlNumberFormat::formatToParts const):
* Source/JavaScriptCore/runtime/IntlPluralRules.cpp:
(JSC::IntlPluralRules::select const):
(JSC::IntlPluralRules::selectRange const):
* Source/JavaScriptCore/runtime/IntlRelativeTimeFormat.cpp:
(JSC::IntlRelativeTimeFormat::formatInternal const):
(JSC::IntlRelativeTimeFormat::formatToParts const):
* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::timeZoneCacheSlow):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::toLocaleCase):
* Source/WTF/wtf/text/StringView.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::subprotocol const):
(WebCore::ThreadableWebSocketChannelClientWrapper::extensions const):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::ExternalCharacterTokenBuffer::ExternalCharacterTokenBuffer):
(WebCore::HTMLTreeBuilder::processTokenInForeignContent):
* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;String&gt;::decode):
(IPC::decodeStringText): Deleted.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a8542ef43120d556721c573faff4085ea0446f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14743 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38527 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114880 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175024 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109554 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5942 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97919 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114693 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111406 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12283 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95262 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39762 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94147 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26901 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81478 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95383 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8011 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28254 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93551 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5794 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8507 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4840 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30411 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14128 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47802 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/102264 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10060 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25499 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
```diff
@@ -105,7 +105,7 @@ static inline std::optional<String> decodeStringText(Decoder& decoder, unsigned
     if (!data.data())
         return std::nullopt;
-    return std::make_optional<String>(data);
+    return std::make_optional<String>(String::fromSpan(data));
```
Should not need the `make_optional` any more after this change.

zdobersek replied:
OK.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -58,14 +58,16 @@ class String final {
     ALWAYS_INLINE static String fromLatin1(const char* characters) { return String { characters }; }
 
+    // Construct a string from a span object.
+    ALWAYS_INLINE static String fromSpan(Span<const LChar> span) { return String { span.data(), static_cast<unsigned>(span.size()) }; }
```
This should be kept grouped in with the Latin-1 constructors and should have Latin1 in its name, since unlike `UChar`, the type `LChar` is not strong enough to firmly establish the character set.

zdobersek replied:
With `fromLatin1Span(Span<const LChar>)`, should UChar variant be `fromUnicodeSpan(Span<const UChar>)`?

darinadler replied:
I don't think so. When the type expresses things unambiguously I think it's good to not make the names so long and explicit. I think `fromSpan` is better.

zdobersek replied:
One downside to separating these two in name is not being able to use them as overloaded functions in templates where the character type is the template parameter. `ArgumentCoder<String>::decode()` and its `decodeStringText` helper are one current example.

It would be possible to constexpr-evaluate the character type and use the appropriate named constructor, but it's not necessarily elegant.

fujii replied:
This is not a problem. The encoder is switching to use `characters8()` or `characters16()`. It's natural that the decoder is also switching.
https://github.com/WebKit/WebKit/blob/4f6caeb7a6616e3ab6f4ea54a9ce483ad535386e/Source/WebKit/Platform/IPC/ArgumentCoders.cpp#L92-L95

zdobersek replied:
Comparative point of 8-bit branching is where the different `decodeStringText` specialization is selected. In this case reworking that codepath and removing the `decodeStringText` template altogether will probably keep things simple enough.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -58,14 +58,16 @@ class String final {
 
+    // Construct a string from a span object.
+    ALWAYS_INLINE static String fromSpan(Span<const LChar> span) { return String { span.data(), static_cast<unsigned>(span.size()) }; }
+    ALWAYS_INLINE static String fromSpan(Span<const UChar> span) { return String { span.data(), static_cast<unsigned>(span.size()) }; }
```
This should be kept grouped in with the other UTF-16 constructors as before.

zdobersek replied:
OK.

darinadler commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -68,11 +68,12 @@ class StringView final {
 
     ALWAYS_INLINE static StringView fromLatin1(const char* characters) { return StringView { characters }; }
 
+    ALWAYS_INLINE static StringView fromSpan(Span<const LChar> span) { return StringView { span.data(), static_cast<unsigned>(span.size()) }; }
```
It would be better to organize and comment these more like the `String` ones, which are clear on the character types and thus seem organized. And this one should be named `fromLatin1Span`.

zdobersek replied:
OK.

darinadler commented on Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
```diff
@@ -100,7 +100,7 @@ String ThreadableWebSocketChannelClientWrapper::subprotocol() const
 {
     if (m_subprotocol.isEmpty())
```
This special case for empty string should not be necessary.

zdobersek replied:
In case of an empty Vector, its data pointer can also be null. So the implicit Span will carry a null data pointer as well, which will end up constructing a null string instead of an empty one, so that would enact a difference in behavior here.

darinadler replied:
OK

darinadler commented on Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
```diff
@@ -114,7 +114,7 @@ String ThreadableWebSocketChannelClientWrapper::extensions() const
 {
     if (m_extensions.isEmpty())
```
This special case for empty string should not be necessary.

fujii commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -58,14 +58,16 @@ class String final {
 
     // Construct a string with Latin-1 data.
     WTF_EXPORT_PRIVATE String(const LChar* characters, unsigned length);
     WTF_EXPORT_PRIVATE String(const char* characters, unsigned length);
```
If I understand the String design policy correctly, we will have to convert these 8bit constructors to `fromLatin1`.
~~~
fromLatin1(const LChar* characters, unsigned length);
fromLatin1(const char* characters, unsigned length);
~~~

If so, I think `fromLatin` should be able to accept `Span<const LChar>`
~~~
fromLatin(Span<const LChar> span)
~~~

darinadler replied:
Maybe, but then we could have the same ambiguity problem again: if you call `String::fromLatin1` on a literal it could use the length of the array rather than the length of the literal. Just depends on how likely we think that mistake is. Requiring the word "span" might help.

zdobersek replied:
Yes, the `String(const CharType*, unsigned)` constructor parameters are effectively spans, so they could all be converted into named constructors taking in Span objects. It's a hefty job, not impossible. Loss of the `unsigned` type for length wouldn't be great, but I think it could be implemented properly.

fujii replied:
We have `fromLatin1(const char*)`. `String::fromLatin1("string")` calls it. No ambiguity problem in this case.

fujii replied:
I attached a patch for the discussion. https://bugs.webkit.org/show_bug.cgi?id=251385#c2

zdobersek committed:
Use named constructors to construct String, StringView from Span objects
https://bugs.webkit.org/show_bug.cgi?id=251385

Reviewed by NOBODY (OOPS!).

Turn conversion constructors on String and StringView classes accepting Span
objects into named constructor equivalents. This avoids unintentional use of
constructors through implicit construction of Span objects from e.g. containers
like Vectors, C arrays or string literals.

For latin-1 data spans, fromLatin1Span() is provided. For UTF-16 data spans, the
generic fromSpan() is provided. StringView constructors are regrouped and
reordered to mirror grouping and ordering of String constructors.

* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::canonicalizeTimeZoneName):
(JSC::IntlDateTimeFormat::format const):
(JSC::IntlDateTimeFormat::createDateIntervalFormatIfNecessary):
(JSC::IntlDateTimeFormat::formatRange):
* Source/JavaScriptCore/runtime/IntlDisplayNames.cpp:
(JSC::IntlDisplayNames::of const):
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::collectElements):
(JSC::IntlDurationFormat::format const):
* Source/JavaScriptCore/runtime/IntlListFormat.cpp:
(JSC::IntlListFormat::format const):
* Source/JavaScriptCore/runtime/IntlNumberFormat.cpp:
(JSC::IntlNumberFormat::format const):
(JSC::IntlNumberFormat::formatToParts const):
* Source/JavaScriptCore/runtime/IntlPluralRules.cpp:
(JSC::IntlPluralRules::select const):
(JSC::IntlPluralRules::selectRange const):
* Source/JavaScriptCore/runtime/IntlRelativeTimeFormat.cpp:
(JSC::IntlRelativeTimeFormat::formatInternal const):
(JSC::IntlRelativeTimeFormat::formatToParts const):
* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::timeZoneCacheSlow):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::toLocaleCase):
* Source/WTF/wtf/text/StringView.h:
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::subprotocol const):
(WebCore::ThreadableWebSocketChannelClientWrapper::extensions const):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::ExternalCharacterTokenBuffer::ExternalCharacterTokenBuffer):
(WebCore::HTMLTreeBuilder::processTokenInForeignContent):
* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder<String>::decode):
(IPC::decodeStringText): Deleted.

darinadler commented on Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
```diff
@@ -1449,7 +1449,7 @@ UDateIntervalFormat* IntlDateTimeFormat::createDateIntervalFormatIfNecessary(JSG
     }
 
-    dataLogLnIf(IntlDateTimeFormatInternal::verbose, "interval format pattern:(", String(pattern), "),skeleton:(", String(skeleton), ")");
+    dataLogLnIf(IntlDateTimeFormatInternal::verbose, "interval format pattern:(", StringView::fromSpan(pattern), "),skeleton:(", StringView::fromSpan(skeleton), ")");
```
Love this improvement. Hate to allocate and deallocate `String` objects for no good reason.

darinadler reviewed (approved):
I guess there is debate about this. Making `String` allocations explicit will make code more wordy and it = will help us spot where we do unnecessary memory allocation. It's great to expand use of `span` and `StringView` in all the cases where we can avoid allocating a new object.

I had forgotten about this proposal and was just doing the reverse, adding a constructor for making `String` from a UTF-8 span without calling `String::fromUTF8`.

Maybe our goal is best served by naming these constructors just `String::create` in all the cases where the type is unambiguous. It's not a goal to have callers have to specify the character encoding in the create function name; the type can express that.

Another possibility is to leave the constructors in place but make them all `explicit`. This could help us achieve the same goal.
